src/Service/Transifex/TranslationsService.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Service\Transifex;
+
+use App\Exception\InvalidFileException;
+use App\Exception\RegexException;
+use App\Service\Transifex\DTO\TranslationDTO;
+use Mautic\Transifex\Connector\Translations;
+use Mautic\Transifex\Exception\ResponseException;
+use Mautic\Transifex\Promise;
+use Mautic\Transifex\TransifexInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Filesystem\Filesystem;
+
+class TranslationsService
+{
+    public function __construct(
+        private readonly TransifexInterface $transifex,
+        private readonly Filesystem $filesystem,
+        private readonly int $downloadMaxAttempts
+    ) {
+    }
+
+    public function getTranslations(TranslationDTO $translationDTO, LoggerInterface $logger): void
+    {
+        $bundlePath = $translationDTO->translationsDir.'/'.$translationDTO->language.'/'.$translationDTO->bundle;
+        $filePath   = $bundlePath.'/'.$translationDTO->file.'.ini';
+
+        $this->filesystem->mkdir($bundlePath);
+
+        // Set the timestamp on the file so our zip builds are reproducible
+        $this->filesystem->touch($bundlePath, strtotime($translationDTO->lastUpdate));
+        $this->filesystem->touch($filePath, strtotime($translationDTO->lastUpdate));
+
+        for ($attempt = 1; $attempt <= $this->downloadMaxAttempts; ++$attempt) {
+            try {
+                $translations = $this->transifex->getConnector(Translations::class);
+                $response     = $translations->download($translationDTO->slug, $translationDTO->language);
+                $promise      = $this->transifex->getApiConnector()->createPromise($response);
+                $promise->setFilePath($filePath);
+                $this->fulfillPromises($promise, $logger);
+                $this->ensureFileValid($promise->getFilePath());
+                break;
+            } catch (ResponseException $e) {
+                if ($attempt === $this->downloadMaxAttempts) {
+                    $this->outputErrors($logger, $filePath, $e->getMessage());
+                    break;
+                }
+
+                sleep(2 ** $attempt);
+            } catch (InvalidFileException|RegexException $e) {
+                $this->outputErrors($logger, $filePath, $e->getMessage());
+                break;
+            }
+        }
+    }
+
+    private function fulfillPromises(Promise $promise, LoggerInterface $logger): void
+    {
+        $promises = new \SplQueue();
+        $promises->enqueue($promise);
+
+        usleep(500000);
```

Comment:
What is this for?

Reply:
This is mentioned in the docs of mautic/Transifex-API (https://github.com/mautic/Transifex-API/blame/master/README.md#L210).

Reply:
Ha, that's fair :)